java/lance-jni/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+    dataset: BlockingDataset,
+) -> JObject<'local> {
+    let j_dataset = create_java_dataset_object(env);
+    // This block sets a native Rust object (dataset) as a field in the Java object (j_dataset).
+    // Caution: This creates a potential for memory leaks. The Rust object (dataset) is not
+    // automatically garbage-collected by Java, and its memory will not be freed unless
+    // explicitly handled.
+    //
+    // To prevent memory leaks, ensure the following:
+    // 1. The Java object (`j_dataset`) should implement the `java.io.Closeable` interface.
+    // 2. Users of this Java object should be instructed to always use it within a try-with-resources
+    //    statement (or manually call the `close()` method) to ensure that `self.close()` is invoked.
+    unsafe {
+        match env.set_rust_field(&j_dataset, "nativeDatasetHandle", dataset) {
+            Ok(_) => j_dataset,
+            Err(err) => {
+                env.throw_new(
+                    "java/lang/RuntimeException",
+                    format!("Failed to set native handle: {}", err),
+                )
+                .expect("Error throwing exception");
+                JObject::null()
+            }
+        }
+    }
+}
+
+fn create_java_dataset_object<'a>(env: &mut JNIEnv<'a>) -> JObject<'a> {
+    env.new_object("com/lancedb/lance/Dataset", "()V", &[])
+        .expect("Failed to create Java Dataset instance")
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_Dataset_open<'local>(
+    mut env: JNIEnv<'local>,
+    _obj: JObject,
+    path: JString,
+) -> JObject<'local> {
+    let path_str = match jni_helpers::extract_path_str(&mut env, &path) {
+        Ok(value) => value,
+        Err(_) => return JObject::null(),
+    };
+
+    match BlockingDataset::open(&path_str) {
+        Ok(dataset) => attach_native_dataset(&mut env, dataset),
+        Err(err) => {
+            env.throw_new(
+                "java/lang/RuntimeException",
+                format!("Failed to open dataset: {}", err),
+            )
+            .expect("Error throwing exception");
+            JObject::null()
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_Dataset_countRows(
+    mut env: JNIEnv,
+    java_dataset: JObject,
+) -> jint {
+    let dataset_guard =
+        unsafe { env.get_rust_field::<_, _, BlockingDataset>(java_dataset, "nativeDatasetHandle") };
+    match dataset_guard {
+        Ok(dataset) => dataset
+            .count_rows()
+            .expect("Faild to get the row count from dataset's metadata.")
+            as jint,
+        Err(_) => -1,
+    }
+}
+
+#[no_mangle]
+pub extern "system" fn Java_com_lancedb_lance_Dataset_releaseNativeDataset(
+    mut env: JNIEnv,
+    obj: JObject,
+) {
+    unsafe {
+        let dataset: BlockingDataset = env
+            .take_rust_field(obj, "nativeDatasetHandle")
+            .expect("Failed to take native dataset handle");
+        dataset.close()
+    }
```

Comment:
nit: again, try to keep `unsafe` scope limited.
```suggestion
    let dataset: BlockingDataset = unsafe { 
        env
        .take_rust_field(obj, "nativeDatasetHandle")
        .expect("Failed to take native dataset handle")
    };
    dataset.close()
```